File: visitor-comparison/src/main/java/com/edwyn/demo/visitor/java25/review/SubtotalVisitor.java
package com.edwyn.demo.visitor.java25.review;

import com.edwyn.demo.visitor.java25.order.DigitalProduct;
import com.edwyn.demo.visitor.java25.order.Discount;
import com.edwyn.demo.visitor.java25.order.OrderItemVisitor;
import com.edwyn.demo.visitor.java25.order.PhysicalProduct;
import com.edwyn.demo.visitor.java25.order.ServiceLine;

public final class SubtotalVisitor implements OrderItemVisitor<Integer> {
    @Override
    public Integer visitPhysicalProduct(PhysicalProduct item) {
        return item.unitPriceCents() * item.quantity();
    }

    @Override
    public Integer visitDigitalProduct(DigitalProduct item) {
        return item.unitPriceCents() * item.licenseCount();
    }

    @Override
    public Integer visitServiceLine(ServiceLine item) {
        return item.hourlyRateCents() * item.hours();
    }

    @Override
    public Integer visitDiscount(Discount item) {
        return -item.amountCents();
    }
}

File: visitor-comparison/src/main/kotlin/com/edwyn/demo/visitor/kotlin/review/SubtotalVisitor.kt
package com.edwyn.demo.visitor.kotlin.review

import com.edwyn.demo.visitor.kotlin.order.DigitalProduct
import com.edwyn.demo.visitor.kotlin.order.Discount
import com.edwyn.demo.visitor.kotlin.order.OrderItem
import com.edwyn.demo.visitor.kotlin.order.PhysicalProduct
import com.edwyn.demo.visitor.kotlin.order.ServiceLine

class SubtotalVisitor {
    fun visit(item: OrderItem): Int =
        when (item) {
            is PhysicalProduct -> item.unitPriceCents * item.quantity
            is DigitalProduct -> item.unitPriceCents * item.licenseCount
            is ServiceLine -> item.hourlyRateCents * item.hours
            is Discount -> -item.amountCents
        }
}
